query_execution.go

```diff
@@ -0,0 +1,898 @@
+				if !ok {
+					builder.WriteString("null")
+					if i < len(filteredSelectionSet)-1 {
+						builder.WriteString(",")
+					}
+					continue
+				}
+				if field.SelectionSet != nil && len(field.SelectionSet) > 0 {
+					innerBody, err := formatResponseDataRec(schema, field.SelectionSet, fieldData, false)
+					if err != nil {
+						return "", err
+					}
+					builder.WriteString(innerBody)
+				} else {
+					fieldJSON, err := json.Marshal(&fieldData)
+					if err != nil {
+						return "", err
+					}
+
+					builder.Write(fieldJSON)
+				}
+			}
+			if i < len(filteredSelectionSet)-1 {
+				builder.WriteString(",")
+			}
+		}
+		if !insideFragment {
+			builder.WriteString("}")
+		}
+	case []interface{}:
+		builder.WriteString("[")
+		for i, v := range result {
+			innerBody, err := formatResponseDataRec(schema, selectionSet, v, false)
+			if err != nil {
+				return "", err
+			}
+			builder.WriteString(innerBody)
+
+			if i < len(result)-1 {
+				builder.WriteString(",")
+			}
+		}
+		builder.WriteString("]")
+	case []map[string]interface{}:
+		builder.WriteString("[")
+		for i, v := range result {
+			innerBody, err := formatResponseDataRec(schema, selectionSet, v, false)
+			if err != nil {
+				return "", err
+			}
+			builder.WriteString(innerBody)
+
+			if i < len(result)-1 {
+				builder.WriteString(",")
+			}
+		}
+		builder.WriteString("]")
+	}
+
+	return builder.String(), nil
+}
+
+// When formatting the response data, the shape of the selection set has to potentially be modified to more closely resemble the shape
+// of the response. This only happens when running into fragments, there are two cases we need to deal with:
+//   1. the selection set of the target fragment has to be unioned with the selection set at the level for which the target fragment is referenced
+//   2. if the target fragments are an implementation of an abstract type, we need to use the __typename from the response body to check which
+//   implementation was resolved. Any fragments that do not match are dropped from the selection set.
+func unionAndTrimSelectionSet(objectTypename string, schema *ast.Schema, selectionSet ast.SelectionSet) (ast.SelectionSet, error) {
+	return unionAndTrimSelectionSetRec(objectTypename, schema, selectionSet, map[string]bool{})
+}
+
+func unionAndTrimSelectionSetRec(objectTypename string, schema *ast.Schema, selectionSet ast.SelectionSet, seenFields map[string]bool) (ast.SelectionSet, error) {
+	var filteredSelectionSet ast.SelectionSet
+	for _, selection := range selectionSet {
+		switch selection := selection.(type) {
+		case *ast.Field:
+			if seenFields[selection.Alias] {
+				continue
+			}
+			seenFields[selection.Alias] = true
+			filteredSelectionSet = append(filteredSelectionSet, selection)
+		case *ast.InlineFragment:
+			fragment := selection
+			if objectTypename == "" {
+				return nil, errors.New("__typename must have been injected when dealing with fragments, check the planner is doing the right thing")
```

Comment:
```suggestion
				return nil, errors.New("__typename must not be blank")
```
Is the longer error meaningful when the branch is not in development? Perhaps check before it is passed to `unionAndTrimSelectionSet`?